File: Easy/js/twoNumSum.js
/*
	Prompt: Write a function that takes in a non-empty array of distinct integers and an integer representing a target sum. If any two numbers in the input array sum up to the target sum, the function should return them in the array in sorted order. If no two numbers sum up to the target sum, the function should return an empty array. Assume that there will be at most one pair of numbers summing up to the target sum.

	Input: [3, 5, -4, 8, 11, 1, -1, 6] , 10
	Output: [-1, 11]
*/

const twoNumSum = (arr, target) => {
  if (!Array.isArray(arr) || !arr.length || !target) {
    return [];
  }

  const memo = {};
  for (let i = 0; i < arr.length; i++) {
    let diff = target - arr[i];
    if (memo[diff] !== undefined) {
      return [arr[i], diff].sort((a, b) => a - b);
    }
    memo[arr[i]] = i;
  }

  return [];
};

console.log(twoNumSum([3, 5, -4, 8, 11, 1, -1, 6], 10));

const twoNumSumBinSearch = (arr, target) => {
  // sanity check
  if (!Array.isArray(arr) || !arr.length || !target) {
    return [];
  }

  const sorted = [...arr].sort((a, b) => a - b);
  let left = 0;
  let right = arr.length - 1;

  while (left < right) {
    let sum = sorted[left] + sorted[right];
    if (sum < target) {
      left++;
    } else if (sum > target) {
      right--;
    } else {
      return [sorted[left], sorted[right]];
    }
  }
  return [];
};

console.log(twoNumSumBinSearch([3, 5, -4, 8, 11, 1, -1, 6], 10));

File: Easy/python/two_num_sum.py
def two_num_sum(arr, target):
    if not arr or type(arr) is not list or not target:
        return []
    memo = {}
    for num in arr:
        diff = target - num
        if diff in memo:
            return sorted([diff, num])
        memo[num] = True
    return []


print(two_num_sum([3, 5, -4, 8, 11, 1, -1, 6], 10))
print(two_num_sum([], 10))
print(two_num_sum('taco', 10))


def two_num_sum_bin_search(arr, target):
    if not arr or type(arr) is not list or not target:
        return []
    sorted_arr = sorted(list(arr))  # make sorted copy
    left = 0
    right = len(arr) - 1
    while left < right:
        the_sum = sorted_arr[left] + sorted_arr[right]
        if the_sum < target:
            left += 1
        elif the_sum > target:
            right -= 1
        else:
            return [sorted_arr[left], sorted_arr[right]]
    return []


print(two_num_sum_bin_search([3, 5, -4, 8, 11, 1, -1, 6], 10))
print(two_num_sum_bin_search([], 10))
print(two_num_sum_bin_search('taco', 10))
